### ISSUE TITLE
Improve docker build with messages when firing rules

### ISSUE THREAD

**Issue by k3ck3c** [docker/docker#19644]
As pylint for Python, we could improve docker build by integrating hadolint from Lukas martinelli
https://github.com/lukasmartinelli/hadolint
and his online checker
http://hadolint.lukasmartinelli.ch/

docker build would still build an image, but the author could see the mistakes and immediately correct his Dockerfile

**Comment by HackToday** [comment]
Sounds good idea. but question is do we have standard rules and error code for Dockerfile ?

It seems better way is to have a general rules and checkers standardized, then everyone can use it as a reference.

**Comment by thaJeztah** [comment]
I really like that linter, but (as @HackToday mentioned), there's no *single* (coding)standard for Dockerfiles. We do have [some recommendations / best practices](https://github.com/docker/docker/blob/8c1d0e31b40438efd1017535babf8429164a401f/docs/articles/dockerfile_best-practices.md), but enforcing those to people is really out of scope for the core functionality of docker.

Note that there are other linters linked from that project, that use different rules; which illustrates that there's no single "truth" as to how a Dockerfile should look.

So, in my opinion; having functionality to check a Dockerfile for syntax errors *could* be considered (after the [builder has been moved out of the daemon](https://github.com/docker/docker/blob/02e18c9ab31cbb7033ec6efebc12274cd7b50e62/ROADMAP.md#122-builder)), but linters should be an external tool, because they may be too opiniated.

**Comment by bamarni** [comment]
> So, in my opinion; having functionality to check a Dockerfile for syntax errors could be considered

What about also making it the default behaviour (not try to build in case of syntax errors)? It is mainly useful for blocks like this : https://github.com/docker-library/redis/blob/bc3ca850f4a2a2e9e5d4b3b8272e5536b4d21d24/3.0/Dockerfile#L28. Here it is very easy to miss a line continuation or make a typo, and time-consuming to restart (no cache layer can be used).

**Comment by k3ck3c** [comment]
A Dockerfile having several FROM, CMD or ENTRYPOINT commands should at
least get a warning, IMHO

2016-04-28 17:25 GMT+02:00 Bilal Amarni <notifications@github.com>:

**Comment by thaJeztah** [comment]
> A Dockerfile having several FROM, CMD or ENTRYPOINT commands should at least get a warning, IMHO

The problem here is that (besides multiple `FROM` being actually valid), we allow overriding the `CMD` or `ENTRYPOINT`. The builder executes each build-step in a new container, implicitly using the image from the previous step as "FROM". Basically, the two examples below are exactly the same;

**Example 1**

```Dockerfile
FROM alpine
CMD echo "hello world"
CMD echo "goodbye world"
```

**Example 2**

```Dockerfile
FROM alpine
CMD echo "hello world"
```

```bash
docker build -t image1 .
```

```Dockerfile
FROM image1
CMD echo "goodbye world"
```

**Comment by k3ck3c** [comment]
Yes, but in example 1, if the author of the Dockerfile expects the 2 echo
commands to print something he will be disppointed, as only the second will
be seen.
That was my point.

The example 2 can be valid/wanted by the author.

2016-05-10 14:23 GMT+02:00 Sebastiaan van Stijn <notifications@github.com>:

**Comment by LK4D4** [comment]
I think it's valid feature to have.

**Comment by thaJeztah** [comment]
While we don't have a full "linter", we _do_ have some basic _syntax_ validation for Dockerfiles before starting the actual build in 1.13, see https://github.com/docker/docker/pull/26516. Additional enhancements are proposed here; https://github.com/docker/docker/issues/26877 (perform validation before sending the build context).
